### ISSUE TITLE
[css-transforms-2] Syntax of individual transforms should reflect general CSS syntax

### ISSUE THREAD

**Issue by evayde** [w3c/csswg-drafts#7646]
For example: Individual translate is only usable like so: `translate: 0 0 10px`
But `translate` as an individual property should be a shorthand for
`translate-x`, `translate-y`, and `translate-z`
In my opinion, this would reflect the current CSS syntax with other properties, like `overflow`, `background-position`, `scroll-snap-type`, and so on. Otherwise, it feels kind of inconsistent to me.

Or is there any technical reason that we cannot provide that syntax for individual transforms?

Spec: https://www.w3.org/TR/css-transforms-2/#individual-transforms

**Comment by jakearchibald** [comment]
@bramus any idea why `translate-x` isn't a thing?

**Comment by jakearchibald** [comment]
So it could be added in future?

**Comment by evayde** [comment]
As far as I understand it, the reason for creating individual transform properties was so that we can animate stuff without writing everything again, so:

```
div {
  transform: translate(0,100px) scale(1.2) rotate(20deg);
}

div:hover {
  transform: translate(10px,200px) scale(1.2) rotate(20deg);
}
```

Can be written like this:

```
div {
  transform: translate(0,100px) scale(1.2) rotate(20deg);
}

div:hover {
  translate: 10px 200px;
}
```

For the same reasons, we could argue that this would be even better (more granular):

```
div {
  transform: translate(0,100px) scale(1.2) rotate(20deg);
}

div:hover {
  translate-y: 200px;
}
```

Also it is more consistent with how other CSS properties and their shorthand syntax works.

**Comment by bramus** [comment]
In addition to `translate-x` and `translate-y`, would it also make sense to have `translate-inline` and `translate-block`? Same for `scale-*`/`rotate-*`, knowing that for the `*-z` properties there would be no logical counterpart.

**Comment by tabatkins** [comment]
We don't have corresponding functions for those; I'm uncomfortable adding properties that are genuinely unexpressible in 'transform' itself.

**Comment by tabatkins** [comment]
This would be changing the decomposition a little, tho not in a way that's observably different from what we have today.

Currently the three transform properties insert three transform functions into the transform stack. Now they'll insert 7: 3 translates, a rotate, and 3 scales. If you continue to use the (now) shorthands the behavior is identical, since translates and scales are order-independent between themselves.

**Comment by dbaron** [comment]
So I think there's a definitely nontrivial cost to adding these, in particular because it requires adding support for separately animating translate-x, translate-y, and translate-z, etc., on the compositor.

If there are strong use cases for the splitting I think that complexity could be justified, but I don't think we should do this just "because a lot of other CSS properties do it", since I think it has significantly more implementation cost than for many other CSS properties.

**Comment by clshortfuse** [comment]
To note, [[css-logical][css-images] flow-relative gradients](https://github.com/w3c/csswg-drafts/issues/1724)
#1724 is blocked until [Align logical values for <position> with the ones defined in CSS Logical Properties](https://github.com/w3c/csswg-drafts/issues/549)
#549 is resolved as per https://github.com/w3c/csswg-drafts/issues/1724#issuecomment-1561542019

I'm not sure we've resolved how `transform-origin` would work yet.

**Comment by bramus** [comment]
Got [another request from an author](https://twitter.com/judehunterdev/status/1689966939160956928) for this today:

> well, I think the problem with those properties is that they still don't give you individual transforms, e.g. translateX and Y separately... so back to square one I guess.

This was explicitly in response to the existence of the current individual transform properties.

**Comment by bramus** [comment]
Quoting @birtles [on Twitter](https://x.com/brianskold/status/1708807450672148965):

> Maybe we need to introduce translateX and translateY properties 😅

This was sparked by [a tweet](https://x.com/mattgperry/status/1708776364944331054?s=20) by author @mattgperry. For their framework [Framer Motion](https://www.framer.com/motion/), IUC, [they internally pipe the passed in `x`/`y`/etc. options to Custom Properties](https://x.com/mattgperry/status/1708809076707139918), which they then use to build a `transform` themselves. This has the downside of [not being hardware accelerated](https://www.bram.us/2023/02/01/the-gotcha-with-animating-custom-properties/).

With translateX/translateY/etc. this would not be a problem, as Framer Motion could then pass these props straight into the individual transform properties.

As Matthew put it in [a follow-up tweet](https://x.com/mattgperry/status/1708808880740859936):

> Yes it would be amazing to be able to set each axis directly!

**Comment by mattgperry** [comment]
In terms of use-case, thinking from the perspective of gestures and/or velocity-based animations, the very first problem I’ve had to solve with every animation library I’ve written is per-axis transform animation. I was surprised when transform/scale/rotate were implemented that they weren’t per-axis. 

For [Motion One](https://motion.dev), which is based purely on WAAPI, the approach has been via custom properties but of course these aren’t accelerated. 

They actually perform worse than the approach taken in [Framer Motion](https://framer.com/motion). This library constructs a new transform at the end of every animation frame but of course has to run on the main thread. Framer Motion uses WAAPI when values can be hardware accelerated and if transforms could be set per axis, many more animations would be.

**Comment by dbaron** [comment]
So I think there are two things we could reasonably do here:
* add `translate-x` and `translate-y` and `translate-z` properties, and make `translate` a shorthand for them
* add `scale-x` and `scale-y` and `scale-z` properties and make `scale` a shorthand for them

I'm curious how strong the use cases are for each of these things.  It sounds like separate translate properties are quite valuable.  Are there similarly strong use cases for the scale properties?

---

I think its less clear what to do about the rotate properties.  The earlier proposal above was to split into `rotate-angle` and `rotate-axis`, but it's not clear to me how strong the use cases for such a split are.

We can't make the existing `rotate` property a shorthand for `rotate-x`/`rotate-y`/`rotate-z` since it supports a syntax that allows rotations around arbitrary vectors, and non-axis-aligned rotations can't be mapped to separate `rotate-x`/`rotate-y`/`rotate-z` properties.  So if we wanted such properties I think we'd have to make them *additional* properties, and then define the order of operations across all *four* rotation properties.  I'd also note that the order of operations just between `rotate-x`/`rotate-y`/`rotate-z` is not obvious.

**Comment by tabatkins** [comment]
> I'd also note that the order of operations just between rotate-x/rotate-y/rotate-z is not obvious.

Yes, this is why I'd object to `rotate-x`/etc properties (and why we didn't do `skew`/etc properties). The TSR order gets you the least-surprising result; anything else will produce surprises at least some of the time.

**Comment by birtles** [comment]
My suggestion on Twitter was partly in jest. I think many of these use cases can already be addressed using animation composition. What is missing there is a better way to control the composite order of the animations.

**Comment by birtles** [comment]
Here is a demo showing independently animated translateX and translateY on the same element. At least in Firefox, both animations are accelerated: https://codepen.io/birtles/pen/mdaGydp

However, the order of composition for script-generated animations is currently based on the order in which the animations are created.

For libraries, I believe they would like to, for example, add a `translateX` animation after the fact and have it composite in a specific order. In the past we've discussed this and decided we could add methods to adjust an animation's relative composite order when the need arose.

**Comment by mattgperry** [comment]
It would indeed probably be enough to be able to manually set transform order rather than rely on animation creation.

On `rotateX` etc I'm not sure I understand why it's complicated. I've always just enforced X/Y/Z and I don't think I've had a comment or complaint in almost ten years. There is a way to define your own order already, via `transform`.

**Comment by clshortfuse** [comment]
I'm here accessibility/localization moreso than animation.

In RTL languages we don't want to touch `translate-y`. We only care to modify `translate-x`. There's a lot of rule rewriting with `translate` alone.

**Comment by tabatkins** [comment]
> I've always just enforced X - Y - Z and I don't think I've had a comment or complaint in almost ten years. 

If an author writes `.foo { rotate-y: 30deg; }`, it's reasonable to assume they expect it to rotate 30deg around the Y axis, aka the vertical line parallel with the page. But if another class has set `rotate-x: 45deg;`, they won't get that result - it'll rotate around the *45deg-canted Y axis* instead. That's pretty surprising! 

The existing TSR order doesn't result in such surprises - without knowing *anything* about what other transform properties are set, you can safely set any individual one and get the result you'd expect from thinking solely about that property in a vacuum - 'rotate' spins the element around regardless of what 'translate' is, 'translate' moves it the specified distance even if it's scaled or rotated, etc.

I think it's fairly useful to maintain this property if we can, and ask authors to write 'transform' if they do want transforms that are less "separable" in this way. It might not be a property we can maintain forever! But it's one we should discard only with caution, and with significant author benefit justifying it, to overcome the downside of author *confusion*.
